Design/purity-checker.md

```diff
@@ -0,0 +1,107 @@
+## Purity Checker
+
+We want to introduce a notion of validity to our `ExpEmbedding` nodes.
+The first violation of validity would be the existence of an impure `assert` statement.
+ 
+
+Every `ExpEmbedding` implements a new `children` property (`:Sequence<ExpEmbedding>`) and the function `isValid()`. 
+
+---
+
+### Preorder
+
+We define an extension function `ExpEmbedding.preorder()` that traverses the `ExpEmbedding` tree and returns an iterable sequence of the nodes in preorder. 
+
+### checkValidity
+
+To check the validity of the graph, we obtain it using the `preorder` function. We filter all embeddings for which we want to check validity and call `isValid()` on them.
+
+```kotlin
+fun ExpEmbedding.checkValidity(): Boolean =
+    preorder()
+        .filterIsInstance<Assert>()
+        .all { it.isValid() }
```

Comment:
Why not call `isValid` on everything, and then just have the default implementation be a no-op?  This will make it easier to e.g. add support for pre-/postconditions.